compiler/rustc_infer/src/infer/mod.rs

```diff
@@ -684,8 +684,8 @@ impl<'tcx> InferOk<'tcx, ()> {
 }
 
 impl<'tcx> InferCtxt<'tcx> {
-    pub fn dcx(&self) -> DiagCtxtHandle<'tcx> {
-        self.tcx.dcx()
+    pub fn dcx(&self) -> DiagCtxtHandle<'_> {
+        self.tcx.dcx().taintable_handle(&self.tainted_by_errors)
```

Comment:
This is what enables the auto-tainting (as long as one goes through this `dcx`). I'm looking into writing a lint that detects when you invoke `dcx` on something that does not taint, while a tainting thing is available